RFDs/0001-job-worker-service.md

```diff
@@ -0,0 +1,468 @@
+---
+authors: Zeeshan Haque (@zphoenixrises)
+state: review
+---
+
+# RFD 0001 - Job Worker Service
+
+## What
+
+This RFD proposes a prototype job worker service that provides a secure API to run arbitrary Linux processes with fine-grained resource control and real-time output streaming. The service aims to demonstrate a scalable and secure approach to process execution in a multi-tenant environment.
+
+## Why
+
+To showcase a secure and efficient approach to running and managing arbitrary processes in a controlled environment.
+
+## Design Overview
+
+### Components
+
+1. **Library**: A reusable job management library
+2. **API**: A gRPC API server
+3. **Client**: A command-line interface (CLI) client
+
+### Key Features
+
+- Job management (start, stop, query status, get output)
+- Real-time output streaming with support for multiple concurrent clients
+- Resource control using cgroups (CPU, Memory, Disk I/O)
+- Secure communication using mTLS
+- Simple role-based authorization scheme
+- gRPC API for client-server communication
+
+#### Key Considerations 
+- This design is not a distributed system and is only optimized to run on a single linux OS
+- If needed, the components of the library can be split into a distributed microservices framework inthe future
+- Rudimentary monitoring of total resource usage will be available through the the CLI if needed.
+
+## Detailed Design
+
+### 1. Library
+
+#### a. In-memory Job Logger
+- Implement a thread-safe circular buffer to store logs with a configurable maximum size
+- Use a read-write mutex to ensure thread-safety for concurrent read/write operations
+- Implement a pub/sub pattern for real-time log streaming:
```

Comment:
I redesigned the Logger to use goroutines with callbacks for publishing logs instead. In the initial design I was thinking of publishing all the logs from the start of execution first.

In the new design, each subscriber when added, will get the output from the start of the process, and all new ouput as they come. The new design will also allow each subscriber to get logs at their own rate. The subscribers will be (mostly) unaaffected by the other subscribers.

Reply:
I'm not sure I understand the motivation for the complexity and limitations of this API design. I would expect a consumer of the library to want to consume an arbitrary number of bytes on each read, or less if less data is available.

Reply:
The requirement for the library states that it should be able to stream to multiple concurrent clients. So, instead of a pull pattern, I assumed that it needs to be a push pattern, wherein the library pushes data to the clients. Am I missing something?

Reply:
As a consumer of the library, I think I'd expect to be able to stream bytes with a standard io.Reader or io.ReadCloser. These may be individually buffered, or they may pull from shared byte storage.

Reply:
Thanks for that clarification. I modified the logger to return an instance of io.Reader that can be used to read an arbitrary number of bytes pulled from the internal byte buffer.